contrib/ubase/patches/libutil-passwd.patch

```diff
@@ -0,0 +1,46 @@
+Index: ubase/ubase/libutil/passwd.c
+===================================================================
+--- ubase.orig/ubase/libutil/passwd.c	2023-01-15 14:57:19.789609155 +0100
++++ ubase/ubase/libutil/passwd.c	2023-01-15 14:59:18.419605250 +0100
+@@ -4,7 +4,6 @@
+ 
+ #include <errno.h>
+ #include <pwd.h>
+-#include <shadow.h>
+ #include <stdio.h>
+ #include <stdlib.h>
+ #include <string.h>
+@@ -19,8 +18,7 @@
+ int
+ pw_check(const struct passwd *pw, const char *pass)
+ {
+-	char *cryptpass, *p;
+-	struct spwd *spw;
++	const char *cryptpass, *p;
+ 
+ 	p = pw->pw_passwd;
+ 	if (p[0] == '!' || p[0] == '*') {
+@@ -35,23 +33,6 @@
+ 		return 0;
+ 	}
+ 
+-	if (pw->pw_passwd[0] == 'x' && pw->pw_passwd[1] == '\0') {
+-		errno = 0;
+-		spw = getspnam(pw->pw_name);
+-		if (!spw) {
+-			if (errno)
+-				weprintf("getspnam: %s:", pw->pw_name);
+-			else
+-				weprintf("who are you?\n");
+-			return -1;
+-		}
+-		p = spw->sp_pwdp;
```

Comment:
From [getpwent(3)](https://man.netbsd.org/NetBSD-9.0/getpwent.3):

> These functions have been written to ``shadow'' the password file, e.g. allow only certain programs to have access to the encrypted password.  If the process which calls them has an effective uid of 0, the encrypted password will be returned, otherwise, the password field of the returned structure will point to the string `*'.

You can fetch `p` here from `pw->pw_passwd`.

Reply:
I'm not sure I understand what is to do here. I deleted this part of `pw_check` because it handles a case of a shadowed password. But since we don't have password shadowing (e.g `getspnam` function) implemented in Mimiker, I decided to remove this source code fragment.

Reply:
BSDs do have password shadowing but done in a more elegant way than in Linux. `master.passwd` is equivalent to `shadow`. However all the data including encrypted passwords is available through single call to `getpwent`. The implementation chooses between `/etc/passwd` and `/etc/master.passwd` based on `euid`.

Please note that our `getpwent` function is incomplete and lacks support for `/etc/passwd`.

Reply:
It seems to me that the idea here was to simply use `getpwnam()` instead of `getspnam()` - https://github.com/cahirwpz/mimiker/pull/1371/commits/5cc5f83eba2987771a43998e704d3d24e7e1e729.